### PR TITLE
Removed dependency on time and chrono crates

### LINKED ISSUE #1123: Remove chrono crate dependency
We are using the `chrono` crate to get `chrono::Utc::now()` and then transform the result to `timestamp_nanos()` that will later be used to calculate a duration. 
We can do that with `time::get_time()` which will return a `Timespec` containing seconds and nanoseconds. So no further processing will be required. Investigate also if SteadyTIme::now` may be a better option.

The nice thing about this is that we will be removing 3-4 crate dependencies.

### PR DESCRIPTION
**Pull request by georgepisaltu**
Instead of using the time and chrono crates, we implemented some functionality in the 'fc_util' crate. It uses the UNIX syscall 'clock_gettime', which is also what time and chrono used on Linux, so no precision is lost. Where possible, we used 'Instant' from 'std::time' to measure elapsed time, as the time crate was deprecated and it is at least as accurate. However, we could not use 'std::time' everywhere as it doesn't support getting CPU clocks and system wall time.

Issue #, if available: fixes #1123 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: api_server/src/http_service.rs [hunk 1]
```diff
@@ -795,6 +795,11 @@ mod tests {
         );
     }
 
+    #[test]
```
Comment (aghecenco) [inline api_server/src/http_service.rs]: Please move this file to a different commit, as the changes are unrelated to the `chrono` dependency removal.

File: devices/src/legacy/rtc_pl031.rs [hunk 2]
```diff
@@ -8,16 +8,17 @@
...
+
 use std::fmt;
-use std::ops::Add;
 use std::{io, result};
 
 use byteorder::{ByteOrder, LittleEndian};
 
 use crate::BusDevice;
 use logger::{Metric, METRICS};
+use std::time::Instant;
```
Comment (aghecenco) [inline devices/src/legacy/rtc_pl031.rs]: Please group imports together:
* first, imports from `std` (i.e. this line should move under the other `use std::`s)
* then, imports from other crates from crates.io
* last, imports from other crates in Firecracker

File: devices/src/legacy/rtc_pl031.rs [hunk 3]
```diff
@@ -90,8 +91,10 @@ impl RTC {
     }
 
     fn get_time(&self) -> u32 {
-        let ts = self.tick_offset.add(SteadyTime::now() - self.previous_now);
-        ts.sec as u32
+        let ts = (self.tick_offset as i128)
+            + (Instant::now().duration_since(self.previous_now).as_nanos() as i128);
+        //ts.sec as u32
```
Comment (aghecenco) [inline devices/src/legacy/rtc_pl031.rs]: Remove this

File: devices/src/legacy/rtc_pl031.rs [hunk 4]
```diff
@@ -107,8 +110,8 @@ impl RTC {
             }
             RTCLR => {
                 self.load = val;
-                self.previous_now = SteadyTime::now();
-                self.tick_offset = Timespec::new(i64::from(val), 0);
+                self.previous_now = Instant::now();
+                self.tick_offset = i64::from(val) * 1_000_000_000;
```
Comment (aghecenco) [inline devices/src/legacy/rtc_pl031.rs]: `1_000_000_000` shows up several times as a magic value, please declare a `const` for it instead

File: devices/src/virtio/net.rs [hunk 5]
```diff
@@ -860,7 +860,7 @@ impl VirtioDevice for Net {
             let tx_queue = queues.remove(0);
             let rx_queue_evt = queue_evts.remove(0);
             let tx_queue_evt = queue_evts.remove(0);
-            let mut mmds_ns = if self.allow_mmds_requests {
+            let mmds_ns = if self.allow_mmds_requests {
```
Comment (aghecenco) [inline devices/src/virtio/net.rs]: The `mut` removals are also unrelated, please group them in another separate commit (this applies to the following `mut` changes as well)

File: fc_util/src/lib.rs [hunk 6]
```diff
@@ -3,57 +3,180 @@
...
-pub fn timestamp_cycles() -> u64 {
-    #[cfg(target_arch = "x86_64")]
-    // Safe because there's nothing that can go wrong with this call.
-    unsafe {
-        std::arch::x86_64::_rdtsc() as u64
-    }
-    #[cfg(not(target_arch = "x86_64"))]
-    {
-        let mut ts = libc::timespec {
+const SECONDS_TO_NANOS: u64 = 1_000_000_000;
```
Comment (aghecenco) [inline fc_util/src/lib.rs]: Ah here's the `const`. Please replace other uses of the magic number `1_000_000_000` with `SECONDS_TO_NANOS` (might need to make it `pub` for that - in which case don't forget to properly document it)

File: fc_util/src/lib.rs [hunk 7]
```diff
@@ -3,57 +3,180 @@
...
+
+/// Wrapper over libc::clockid_t to specify Linux Kernel clock source.
```
Comment (aghecenco) [inline fc_util/src/lib.rs]:
```suggestion
/// Wrapper over `libc::clockid_t` to specify Linux Kernel clock source.
```

File: fc_util/src/lib.rs [hunk 8]
```diff
@@ -3,57 +3,180 @@
...
+pub enum ClockType {
+    /// Equivalent to libc::CLOCK_MONOTONIC.
```
Comment (aghecenco) [inline fc_util/src/lib.rs]: Please enclose all `libc::...` in backtics (`libc::CLOCK_MONOTONIC` etc)

File: fc_util/src/lib.rs [hunk 9]
```diff
@@ -3,57 +3,180 @@
...
+    /// Months in current year.
+    mon: i32,
+    /// Years passed since 1900 BC.
+    year: i32,
+    /// Nanoseconds in current second.
+    nsec: i64,
+}
+
+impl LocalTime {
+    /// Returns the LocalTime structure for the calling moment.
```
Comment (aghecenco) [inline fc_util/src/lib.rs]: To make this show up nicely in the generated documentation:
```suggestion
    /// Returns the [`LocalTime`](struct.LocalTime.html) structure for the calling moment.
```

File: fc_util/src/lib.rs [hunk 10]
```diff
@@ -3,57 +3,180 @@
...
+    pub fn now() -> LocalTime {
+        let mut a: libc::timespec = libc::timespec {
```
Comment (aghecenco) [inline fc_util/src/lib.rs]: No need to specify the type if you assign a value of that type.
I mean
```suggestion
        let mut a = libc::timespec {
```

File: fc_util/src/lib.rs [hunk 11]
```diff
@@ -3,57 +3,180 @@
...
             tv_sec: 0,
             tv_nsec: 0,
         };
+        let mut b: libc::tm = libc::tm {
```
Comment (aghecenco) [inline fc_util/src/lib.rs]: Same here.

File: fc_util/src/lib.rs [hunk 12]
```diff
@@ -3,57 +3,180 @@
...
+            tm_isdst: 0,
+            tm_gmtoff: 0,
+            tm_zone: std::ptr::null(),
+        };
 
+        // Safe because the parameters are valid.
         unsafe {
-            libc::clock_gettime(libc::CLOCK_MONOTONIC, &mut ts);
+            libc::clock_gettime(libc::CLOCK_REALTIME, &mut a);
+            libc::localtime_r(&a.tv_sec, &mut b);
```
Comment (aghecenco) [inline fc_util/src/lib.rs]: The `unsafe` block should end here.

File: fc_util/src/lib.rs (hunk 11 above)
Comment (aghecenco) [reply fc_util/src/lib.rs]: Ah and please use more meaningful variable names than `a` and `b`.

File: fc_util/src/lib.rs [hunk 13]
```diff
@@ -3,57 +3,180 @@
...
+            self.hour,
+            self.min,
+            self.sec,
+            self.nsec
+        )
+    }
 }
 
-pub fn now_cputime_us() -> u64 {
+pub fn timestamp_cycles() -> u64 {
```
Comment (aghecenco) [inline fc_util/src/lib.rs]: Please add documentation to this `pub fn`.

File: fc_util/src/lib.rs [hunk 14]
```diff
@@ -3,57 +3,180 @@
...
+///
+/// # Arguments
+///
+/// * `clock_type` - Identifier of the Linux Kernel clock on which to act.
+pub fn get_time(clock_type: ClockType) -> u64 {
     let mut time_struct = libc::timespec {
         tv_sec: 0,
         tv_nsec: 0,
     };
+    let clock_type: libc::clockid_t = match clock_type {
```
Comment (aghecenco) [inline fc_util/src/lib.rs]: This would be nicer with an `impl Into<libc::clockid_t> for ClockType`.

File: logger/src/lib.rs [hunk 15]
```diff
@@ -136,17 +136,16 @@
 //! metric will get increased.
 //! Metrics are only logged to pipes. Logs can be flushed either to stdout/stderr or to a pipe.
 
-extern crate chrono;
 // workaround to macro_reexport
 #[macro_use]
 extern crate lazy_static;
+extern crate fc_util;
```
Comment (aghecenco) [inline logger/src/lib.rs]: Please move this `extern crate` after all the crate imports that come from crates.io.

File: logger/src/lib.rs [hunk 16]
```diff
@@ -159,7 +158,7 @@ use std::str::FromStr;
 use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
 use std::sync::{Mutex, MutexGuard, RwLock};
 
-use chrono::Local;
+use fc_util::LocalTime;
```
Comment (aghecenco) [inline logger/src/lib.rs]: Move this `use` lower too., with other `use`s from crates in Firecracker.

File: api_server/src/http_service.rs [hunk 17]
```diff
@@ -795,6 +795,11 @@ mod tests {
...
+    fn test_log_received_api_request() {
```
Comment (andreeaflorescu) [inline api_server/src/http_service.rs]: I am strongly against tests that are only for coverage. If you do want to add a test because coverage is decreased, please write something where the expected result of the function is checked.

File: fc_util/src/lib.rs [hunk 18]
```diff
@@ -3,57 +3,189 @@
...
+        let local_time = LocalTime {
+            sec: 5,
+            min: 5,
+            hour: 5,
+            mday: 23,
+            mon: 8,
+            year: 44,
+            nsec: 123,
+        };
+        assert_eq!(
```
Comment (andreeaflorescu) [inline fc_util/src/lib.rs]: Why not use `assert_eq!(String::from("1944-08-23T05:05:05.000000123"), local_time.to_string());`

File: rate_limiter/src/lib.rs [hunk 19]
```diff
@@ -162,9 +175,9 @@ impl TokenBucket {
             }
         }
         // Compute time passed since last refill/update.
-        let now = time::precise_time_ns();
-        let time_delta = now - self.last_update;
-        self.last_update = now;
+        let time_delta = self.last_update.elapsed().as_secs() * 1_000_000_000
```
Comment (andreeaflorescu) [inline rate_limiter/src/lib.rs]: Shouldn't we use the `fc_util::SECONDS_TO_NANOS` constant here?

File: fc_util/src/lib.rs [hunk 20]
```diff
@@ -3,57 +3,189 @@
...
+
 pub mod validators;
 
-pub fn timestamp_cycles() -> u64 {
-    #[cfg(target_arch = "x86_64")]
-    // Safe because there's nothing that can go wrong with this call.
-    unsafe {
-        std::arch::x86_64::_rdtsc() as u64
+/// Constant to convert seconds to nanoseconds.
+pub const SECONDS_TO_NANOS: u64 = 1_000_000_000;
```
Comment (andreeaflorescu) [inline fc_util/src/lib.rs]: `SECONDS_TO_NANOS` is not the best name for that constant. Can you change it to something that makes more sense?

Examples:
- `SEC_TO_NANO_MULTIPLIER`
- `SECOND_TO_NANOSECOND`
- `SEC_TO_NANO`

Basically I would name it so that it's clear that it's a constant which represents the multiplier for converting one second to one nanosecond.

Better yet, you can create a public function: `seconds_to_nanoseconds` which would farther improve the readability of the code.

```
// In fc_util:
pub fn seconds_to_nanoseconds(value: i64) -> Option<i64> {
        value.checked_mul(1_000_000_000)
}

// Code that was using the constant would look like this:
self.tick_offset = seconds_to_nanoseconds(i64::from(val)).unwrap();
// In some cases you might want to check the value of seconds_to_nanoseconds. 
// In RTC if that value overflows we most likely have a malicious user and we want to terminate
// the execution of Firecracker so we use unwrap.

```

Comment (acatangiu) [reply fc_util/src/lib.rs]: `NANOS_PER_SECOND` 🐎

File: logger/src/lib.rs [hunk 21]
```diff
@@ -786,7 +783,7 @@ impl Log for Logger {
         if self.enabled(record.metadata()) {
             let msg = format!(
                 "{}{}{}{}",
-                Local::now().format(TIME_FMT),
+                LocalTime::now(),
```
Comment (andreeaflorescu) [inline logger/src/lib.rs]: Is the format of the log entries changed? If yes, than you should also update the CHANGELOG.md file and specify that the format is changed.

Comment (georgepisaltu) [reply logger/src/lib.rs]: The format is the same

File: devices/src/legacy/rtc_pl031.rs [hunk 22]
```diff
@@ -203,18 +205,17 @@ mod tests {
         assert_eq!(v, 123);
 
         // Read and write to the LR register.
-        let v = time::get_time();
-        LittleEndian::write_u32(&mut data, v.sec as u32);
+        let v = fc_util::get_time(fc_util::ClockType::Real);
```
Comment (andreeaflorescu) [inline devices/src/legacy/rtc_pl031.rs]: Instead of writing fc_util::something, please import what you need from fc_util.

File: rate_limiter/src/lib.rs [hunk 23]
```diff
@@ -95,13 +95,27 @@ pub struct TokenBucket {
...
     budget: u64,
-    last_update: u64,
+    last_update: Instant,
 
     // Fields used for pre-processing optimizations.
     processed_capacity: u64,
     processed_refill_time: u64,
 }
 
+impl Default for TokenBucket {
```
Comment (andreeaflorescu) [inline rate_limiter/src/lib.rs]: The default implementation for TokenBucket is not used anywhere. Please remove it in a different commit before the commit that replaces chrono and time with the custom implementation from fc_util.

File: devices/src/legacy/rtc_pl031.rs [hunk 24]
```diff
@@ -107,8 +109,8 @@ impl RTC {
             }
             RTCLR => {
                 self.load = val;
-                self.previous_now = SteadyTime::now();
-                self.tick_offset = Timespec::new(i64::from(val), 0);
+                self.previous_now = Instant::now();
+                self.tick_offset = fc_util::seconds_to_nanoseconds(i64::from(val)).unwrap();
```
Comment (andreeaflorescu) [inline devices/src/legacy/rtc_pl031.rs]: Add a comment about why this unwrap is safe. Or specifically when this unwrap is unsafe i.e. when you have a malicious guest in which case you want to terminate the execution of firecracker.
